client/rest-high-level/src/main/java/org/elasticsearch/client/security/PutUserRequest.java

```diff
@@ -0,0 +1,155 @@
+    private final RefreshPolicy refreshPolicy;
+
+    public PutUserRequest(String username, char[] password, List<String> roles, String fullName, String email, boolean enabled,
+                          Map<String, Object> metadata, RefreshPolicy refreshPolicy) {
+        this.username = Objects.requireNonNull(username, "username is required");
+        this.password = password;
+        this.roles = Collections.unmodifiableList(Objects.requireNonNull(roles, "roles must be specified"));
+        this.fullName = fullName;
+        this.email = email;
+        this.enabled = enabled;
+        this.metadata = metadata == null ? Collections.emptyMap() : Collections.unmodifiableMap(metadata);
+        this.refreshPolicy = refreshPolicy == null ? RefreshPolicy.getDefault() : refreshPolicy;
+    }
+
+    public String getUsername() {
+        return username;
+    }
+
+    public List<String> getRoles() {
+        return roles;
+    }
+
+    public String getFullName() {
+        return fullName;
+    }
+
+    public String getEmail() {
+        return email;
+    }
+
+    public Map<String, Object> getMetadata() {
+        return metadata;
+    }
+
+    public char[] getPassword() {
+        return password;
+    }
+
+    public boolean isEnabled() {
+        return enabled;
+    }
+
+    public RefreshPolicy getRefreshPolicy() {
+        return refreshPolicy;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        PutUserRequest that = (PutUserRequest) o;
+        return enabled == that.enabled &&
+            Objects.equals(username, that.username) &&
+            Objects.equals(roles, that.roles) &&
+            Objects.equals(fullName, that.fullName) &&
+            Objects.equals(email, that.email) &&
+            Objects.equals(metadata, that.metadata) &&
+            Arrays.equals(password, that.password) &&
+            refreshPolicy == that.refreshPolicy;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = Objects.hash(username, roles, fullName, email, metadata, enabled, refreshPolicy);
+        result = 31 * result + Arrays.hashCode(password);
+        return result;
+    }
+
+    @Override
+    public void close() {
+        if (password != null) {
+            Arrays.fill(password, (char) 0);
+        }
+    }
+
+    @Override
+    public Optional<ValidationException> validate() {
+        if (metadata != null && metadata.keySet().stream().anyMatch(s -> s.startsWith("_"))) {
+            ValidationException validationException = new ValidationException();
+            validationException.addValidationError("metadata keys may not start with [_]");
+            return Optional.of(validationException);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
```

Comment:
Are we missing username here ?

Reply:
> Looks like the action expects it as request param just like refresh policy

I think there's a confusion here. `username` is a URI path component parameter that ends up in the same `map` as request parameters (`refresh`).

I like it like it is right now, I would not add the username to the request body since we use the PUT method and the resource name (username) is part of the URL. IMO adding it to the body aswell would be redundant.

Reply:
for 7.0, we need to examine what the current REST API sends back and replicate it. If thats sans user, we can add it if we want, but if it has a user in the body, we need to leave it for 7.0

Reply:
We return username in the body for get users so I'll add it to this put body for consistency; it is optional and all that happens is we validate that the url path for username matches the username in the body.